MetadataExtractor/IO/RandomAccessStream.cs

```diff
@@ -0,0 +1,561 @@
+﻿// Copyright (c) Drew Noakes and contributors. All Rights Reserved. Licensed under the Apache License, Version 2.0. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using JetBrains.Annotations;
+
+namespace MetadataExtractor.IO
+{
+    /// <summary>Reads and buffers data in chunks and provides methods for reading data types</summary>
+    /// <remarks>
+    /// This class implements buffered reading of data typically for use with <see cref="ReaderInfo"/> 
+    /// objects and provides methods for reading data types from it. Data is captured in configurable 
+    /// chunks for efficiency. Both seekable and non-seekable streams are supported.
+    /// </remarks>
+    /// <author>Kevin Mott https://github.com/kwhopper</author>
+    /// <author>Drew Noakes https://drewnoakes.com</author>
+    public class RandomAccessStream
+    {
+        private Stream? p_inputStream;
+        private long p_streamLength = -1;
+
+        //private readonly List<ReaderInfo> rdrList = new List<ReaderInfo>();
+        private bool p_isStreamFinished;
+
+        private const int DefaultChunkLength = 4 * 1024;
+        private readonly int p_chunkLength;
+        public Dictionary<long, byte[]> p_chunks = new Dictionary<long, byte[]>();
+
+        public RandomAccessStream(Stream stream, long streamLength = -1)
+        {
+            if (stream == null)
+                throw new ArgumentNullException(nameof(stream));
+
+            if (streamLength == -1)
+                streamLength = stream.CanSeek ? stream.Length : int.MaxValue;
```

Comment:
I inferred that from the "Remarks" section here:

[https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.canseek?view=netframework-4.5](https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.canseek?view=netframework-4.5)

This seems like a reasonable assumption, particularly with Microsoft-provided classes like FileStream. If CanSeek=true but the length is unknown, CanSeek wouldn't provide a lot of new information to the caller. I think MS is trying to say that CanSeek implies certain other things, with one being Length. Let me know if you have other thoughts.

It's certainly possible for someone to inherit from Stream, allow CanSeek to be true, and purposely throw exceptions on Length. An implementation of that sort doesn't seem very useful.

Reply:
Interesting. I am surprised by that. Given the members are virtual, there could be other subclasses that don't support this policy. However it looks promising.

The docs on `Length` don't say that it won't throw when `CanSeek` is `true`. I'd suggest catching `NotSupportedException` and setting `MaxValue` in case of exception.

Reply:
It's now wrapped in a try/catch